### PR TITLE
#8282: Support non-4d tensor and fp32_dest_acc_en for moreh nllloss backward

### PR DESCRIPTION
**Pull request by hschoi4448**
- refactoring moreh nllloss backward
  - support non 4d tensor input
  - fp32_dest_acc_en

- Add moreh helper functions for fp32_dest_acc_en
  - refactoring moreh_nll_loss forward kernels

### CODE REVIEW

File: tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/moreh_nll_loss_backward_kernel.cpp [hunk 1]
```diff
@@ -4,99 +4,110 @@
...
-
-#include "compute_kernel_api/bcast.h"
-#include "compute_kernel_api/eltwise_binary.h"
-#include "compute_kernel_api/eltwise_unary/negative.h"
-#include "compute_kernel_api/eltwise_unary/recip.h"
-#include "compute_kernel_api/tile_move_copy.h"
-
-ALWI void ACQ() { acquire_dst(tt::DstMode::Half); }
-ALWI void REL() { release_dst(tt::DstMode::Half); }
+#include "debug/dprint.h"  // required in all kernels using DPRINT
```
Comment (seunghwan100) [inline tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/moreh_nll_loss_backward_kernel.cpp:7]: I think it's better to remove this.

File: tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        .page_size = output_grad_tile_bytes,
+        .data_format = output_grad_data_format};
+    constexpr uint32_t onetile = 1;
+
+    union {
+        float f;
+        uint32_t u;
+    } one, zero;
+    one.f = 1.0f;
+    zero.f = 0.0f;
```
Comment (seunghwan100) [inline tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp:58]: How about using this?
https://github.com/tenstorrent/tt-metal/blob/590d0c2fa142ca0047061366219ec88f740bde11/tt_eager/tt_dnn/kernels/dataflow/moreh_common.hpp#L41C1-L44C3

File: tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/moreh_nll_loss_backward_kernel.cpp (hunk 1 above)
Comment (hschoi4448) [reply tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/moreh_nll_loss_backward_kernel.cpp]: removed the dprint header.

File: tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp (hunk 2 above)
Comment (hschoi4448) [reply tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp]: Now use the `Scalar` structure

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_nll_loss.py [hunk 3]
```diff
@@ -386,22 +324,23 @@ def test_moreh_nll_loss_2d_backward(shape, ignore_index, reduction_mean, device,
...
+    for _ in range(2):
+        tt_input_grad = ttl.operations.primary.moreh_nll_loss_backward(
+            tt_target,
+            tt_weight,
+            tt_divisor,
+            tt_output_grad,
+            tt_input_grad,
+            ignore_index,
+            reduction_mean,
+        )
```
Comment (TT-BrianLiu) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_nll_loss.py:342]: What is the point of this loop? If it's too test program cache:
- add an assert on program entries
- shift the input/output memory by adding some dummy tensor

File: tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    const InterleavedAddrGenFast<output_grad_is_dram> addrg_output_grad = {
+        .bank_base_address = output_grad_addr,
+        .page_size = output_grad_tile_bytes,
+        .data_format = output_grad_data_format};
+    constexpr uint32_t onetile = 1;
+
+#if defined(WEIGHT)
+    const InterleavedAddrGen<weight_is_dram> addrg_weight = {
+        .bank_base_address = weight_addr,
+        .page_size = 1024 * element_size,
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp:55]: You can get tile size from CB with `get_tile_size`

File: tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    noc_async_read_barrier();
+    cb_push_back(cb_output_grad, onetile);
+
+
+    uint32_t Ct = (C + TILE_HEIGHT - 1) / TILE_HEIGHT;
+
+    uint32_t end_id = start_id + num_tiles_per_core;
+    for (uint32_t i = start_id; i < end_id; ++i) {
+        uint32_t nt = i / Ct;
+        int32_t ct = static_cast<int32_t>(i % Ct);
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp:98]: Why int and not uint?

Comment (hschoi4448) [reply tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp]: It was a mistake to use int32_t. It has been corrected to uint32_t.

File: tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp (hunk 4 above)
Comment (hschoi4448) [reply tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp]: It has been modified to use `get_tile_size`

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_nll_loss.py (hunk 3 above)
Comment (hschoi4448) [reply tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_nll_loss.py]: > What is the point of this loop? If it's to test program cache:

Yes, this is to test the program cache. If it runs only once, the code related to override_runtime_args_callback will not be executed.

> add an assert on program entries

I didn't understand the previous statement. 
There are asserts like:
```C++
            TT_ASSERT(input_tensors.size() == 2);
            TT_ASSERT(optional_input_tensors.size() == 2);
            TT_ASSERT(output_tensors.size() == 1);
```
 If not, can you provide an example?

>shift the input/output memory by adding some dummy tensor

I've modified the callback test to receive random input every time.

File: tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp (hunk 5 above)
Comment (TT-BrianLiu) [reply tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp]: There are still other instances of this I think. Can you clean those up as well?

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_nll_loss.py (hunk 3 above)
Comment (TT-BrianLiu) [reply tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_nll_loss.py]: If you want to test program cache:
- _Ensure that program cache is actually hit._ You do this by checking that the number of generated caches for an op is 1 (or however many your test generates) when you loop it twice. You can query the number of program caches with `device.num_program_cache_entries()`. 
- _Ensure that the callback is actually correct._ To properly test the callback, you have to make sure that the runtime arg updates actually matter in your test. Most often, the callback updates the inputs/output buffer addresses. Trivially looping the test most often results in your input/output tensors being in the same location for two runs (ie. same buffer addresses). If this is the case, the test will always pass when you don't have anything in the callback, even if your data is different the second time. To convince yourself, you can print out the value of the args being updated in the callback and see if it's the same as the first time the program was compiled and launched. If it is, then the test isn't really testing anything. One hack is to shift the device memory of where your inputs/output are expected to be by allocating a small tensor in between the loops (see below). As an exercise, you can again comment out the callback and now you will see the test actually fail the second time since your inputs/output are actually in a different location now.

image

Comment (hschoi4448) [reply tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_nll_loss.py]: >Trivially looping the test most often results in your input/output tensors being in the same location for two runs (ie. same buffer addresses).

Thank you for the comment. As you mentioned, by running the for loop twice without tensor allocation, I confirmed that the tensor addresses remained the same. Therefore, I moved the tensor creation inside the for loop, and I observed that the addresses were updated. Since the buffer addresses change without a dummy tensor, it seems the dummy tensor is unnecessary.

This has benn fixed in 'https://github.com/tenstorrent/tt-metal/pull/8966/commits/f3c5ab9765d96881f2c29363261db3a059103d78'

```Python
    for _ in range(2):
        # In each loop, a new tt tensor and value are created.
        (torch_input, torch_target, torch_weight, torch_divisor, torch_output) = get_torch_tensors(shape)
        if none_weight:
            torch_weight = None

        (tt_input, tt_target, tt_weight, tt_divisor, tt_output) = get_tt_tensors(
            torch_input, torch_target, torch_weight, torch_divisor, torch_output, device
        )

        tt_loss = ttl.operations.primary.moreh_nll_loss(
            tt_input,
            tt_target,
            tt_weight,
            tt_divisor,
            tt_output,
            ignore_index,
            reduction_mean,
        )
```


> device.num_program_cache_entries().

The above function appears to print the number of generated caches. 
However, in the case of `NLL loss`, it internally calls `moreh_sum`, and the number of generated caches depends on the implementation of `moreh_sum`. Therefore, checking this number might not be the correct approach.

Instead of the `num_program_cache_entries` function, how about adding a boolean variable to `ProgramCache` along with `enable_cache_check()` and `disable_cache_check()` functions, and then incorporating checks like `TT_ASSERT(cache_hit)`?

```C++
struct ProgramCache {
    inline std::optional<std::shared_ptr<void>> find(uint64_t program_hash) {
        auto cache_hit = this->cache_.count(program_hash) > 0;
        if (is_cache_check_enabled_ ) {
           TT_ASSERT(cache_hit);
        }
        if (cache_hit) {
            return this->cache_.at(program_hash);
        }
        return std::nullopt;
    }

    void enable_cache_check() {
        is_cache_check_enabled_ = true;
    }

    void disable_cache_check() {
        is_cache_check_enabled_ = false;
    }

  private:
    inline static bool is_cache_check_enabled_ = false;
}
```

File: tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp (hunk 5 above)
Comment (hschoi4448) [reply tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp]: Yes,  I've cleaned up similar cases in this commit.
```C++
        int32_t c = static_cast<int32_t>(nc % C);
->
        uint32_t c = nc % C;
```


https://github.com/tenstorrent/tt-metal/pull/8966/commits/373a9fed99806c2bc4be03a588ba987530a26d42#diff-2a1285f3fc0f375940074cef369b396765945f23f0070a3865001b0d758d6f87L99

**Review by TT-BrianLiu — APPROVED** [review]
